File: frontend/src/types/player.ts

export type Position = "PG" | "SG" | "SF" | "PF" | "C";

export type ShotZoneId =
  | "restricted_area"
  | "paint_non_ra"
  | "midrange_left_baseline"
  | "midrange_left_wing"
  | "midrange_center"
  | "midrange_right_wing"
  | "midrange_right_baseline"
  | "three_left_corner"
  | "three_left_wing"
  | "three_center"
  | "three_right_wing"
  | "three_right_corner"
  | "backcourt"
  | "post_up";

export type InjuryType =
  | "sprain"
  | "strain"
  | "fracture"
  | "tear"
  | "contusion"
  | "concussion"
  | "soreness"
  | "inflammation"
  | "dislocation";

export type BodyPart =
  | "ankle"
  | "knee"
  | "shoulder"
  | "back"
  | "wrist"
  | "hand"
  | "foot"
  | "hamstring"
  | "groin"
  | "calf"
  | "head"
  | "hip"
  | "quad"
  | "elbow"
  | "finger"
  | "thumb"
  | "rib"
  | "achilles";

export interface PlayerBio {
  firstName: string;
  lastName: string;
  position: Position;
  secondaryPosition: Position | null;
  height: number;
  weight: number;
  age: number;
  yearsInLeague: number;
  college: string | null;
  country: string;
  draftYear: number;
  draftRound: number;
  draftPick: number;
  jerseyNumber: number;
  hand: "L" | "R";
}

export interface PlayerRatings {
  finishing: number;
  closeRange: number;
  midRange: number;
  threePoint: number;
  freeThrow: number;
  postGame: number;
  drawFoul: number;
  offBallMovement: number;
  ballHandling: number;
  passingVision: number;
  passingAccuracy: number;
  perimeterDefense: number;
  interiorDefense: number;
  shotBlocking: number;
  stealing: number;
  defensiveIq: number;
  defensiveConsistency: number;
  speed: number;
  acceleration: number;
  lateralQuickness: number;
  vertical: number;
  strength: number;
  stamina: number;
  basketballIq: number;
  offensiveIq: number;
  rebounding: number;
  offensiveRebounding: number;
  hustle: number;
  intangibles: number;
  overall: number;
  /** Skill-derived overall before form and injury adjustments. */
  baseOverall?: number;
  potential: number;
  peakAge: number;
}

export interface ShotZone {
  zoneId: ShotZoneId;
  tendency: number;
  makeRate: number;
}

export interface ShotChartProfile {
  zones: ShotZone[];
}

export interface PlayerTendencies {
  pullUpFrequency: number;
  catchAndShootFrequency: number;
  driveFrequency: number;
  postUpFrequency: number;
  isoFrequency: number;
  pickAndRollBallHandler: number;
  pickAndRollScreener: number;
  spotUpFrequency: number;
  transitionFrequency: number;
  cutFrequency: number;
  passOutOfDriveRate: number;
  skipPassRate: number;
  alleyOopPassRate: number;
  gambleForSteals: number;
  helpDefenseRate: number;
  closeoutAggression: number;
  boxOutRate: number;
  usageDesire: number;
  pacePreference: number;
  foulProneness: number;
  shotClockTendency: number;
  contestedShotWillingness: number;
}

export interface CharacterTraits {
  leadership: number;
  workEthic: number;
  clutch: number;
  ego: number;
  coachability: number;
  temperament: number;
  fanFavorite: number;
  mediaPersonality: number;
  loyalty: number;
  competitiveness: number;
}

export interface InjuryRecord {
  type: InjuryType;
  severity: "minor" | "moderate" | "severe" | "season_ending";
  gamesOut: number;
  seasonYear: number;
  bodyPart: BodyPart;
  /** True when the injury permanently reduced physical ratings (e.g. an ACL tear). */
  permanentEffect?: boolean;
}

export interface DurabilityProfile {
  overallDurability: number;
  ankleHealth: number;
  kneeHealth: number;
  shoulderHealth: number;
  backHealth: number;
  wristHandHealth: number;
  footHealth: number;
  concussionRisk: number;
  softTissueRisk: number;
  injuryHistory: InjuryRecord[];
}

export interface ActiveInjury {
  bodyPart: BodyPart;
  type: InjuryType;
  severity: "minor" | "moderate" | "severe" | "season_ending";
  gamesRemaining: number;
  dateInjured: string;
  /** Coach elected to keep the player on the floor at reduced effectiveness. */
  playingThrough?: boolean;
}

export interface PlayerStatus {
  health: "healthy" | "day_to_day" | "out" | "injured_reserve";
  currentInjury: ActiveInjury | null;
  fatigue: number;
  morale: number;
  isRookie: boolean;
  isFreeAgent: boolean;
  isRestrictedFA: boolean;
  teamId: string | null;
  /** Hot/cold streak adjustment shown in the displayed overall (-2..+2). */
  form?: number;
  /** Smoothed performance-vs-norm signal that drives form. */
  formMomentum?: number;
}

export interface SeasonStats {
  season: string;
  team: string;
  gp: number;
  gs: number;
  mpg: number;
  ppg: number;
  rpg: number;
  apg: number;
  spg: number;
  bpg: number;
  topg: number;
  fgm: number;
  fga: number;
  fg_pct: number;
  three_pm: number;
  three_pa: number;
  three_pct: number;
  ftm: number;
  fta: number;
  ft_pct: number;
  orpg: number;
  drpg: number;
  pfpg: number;
}

export interface PlayerContract {
  annualSalary: number;
  yearsRemaining: number;
  totalYears: number;
  contractType: string;
  noTradeClause: boolean;
  playerOption: boolean;
  teamOption: boolean;
  guaranteed: boolean;
}

export interface Player {
  id: string;
  nbaId: number;
  headshotUrl: string;
  teamId: string;
  bio: PlayerBio;
  ratings: PlayerRatings;
  shotChart: ShotChartProfile;
  tendencies: PlayerTendencies;
  character: CharacterTraits;
  durability: DurabilityProfile;
  contract: PlayerContract | null;
  status: PlayerStatus;
  careerStats: SeasonStats[];
  awards: string[];
}

File: engine/models/player.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class InjuryRecord:
    injury_type: str
    severity: str
    body_part: str
    games_out: int
    season_year: int


@dataclass
class ActiveInjury:
    body_part: str
    injury_type: str
    severity: str
    games_remaining: int
    date_injured: str


@dataclass
class PlayerBio:
    first_name: str
    last_name: str
    position: str
    secondary_position: str | None
    height: int
    weight: int
    age: int
    years_in_league: int
    college: str | None
    country: str
    draft_year: int
    draft_round: int
    draft_pick: int
    jersey_number: int
    hand: str


@dataclass
class PlayerRatings:
    finishing: int
    close_range: int
    mid_range: int
    three_point: int
    free_throw: int
    post_game: int
    draw_foul: int
    off_ball_movement: int
    ball_handling: int
    passing_vision: int
    passing_accuracy: int

    perimeter_defense: int
    interior_defense: int
    shot_blocking: int
    stealing: int
    defensive_iq: int
    defensive_consistency: int

    speed: int
    acceleration: int
    lateral_quickness: int
    vertical: int
    strength: int
    stamina: int

    basketball_iq: int
    offensive_iq: int
    rebounding: int
    offensive_rebounding: int
    hustle: int

    overall: int
    potential: int
    peak_age: int


@dataclass
class ShotZone:
    zone_id: str
    tendency: float
    make_rate: float


@dataclass
class ShotChartProfile:
    zones: list[ShotZone]

    def get_zone(self, zone_id: str) -> ShotZone | None:
        for zone in self.zones:
            if zone.zone_id == zone_id:
                return zone
        return None


@dataclass
class PlayerTendencies:
    pull_up_frequency: int
    catch_and_shoot_frequency: int
    drive_frequency: int
    post_up_frequency: int
    iso_frequency: int
    pick_and_roll_ball_handler: int
    pick_and_roll_screener: int
    spot_up_frequency: int
    transition_frequency: int
    cut_frequency: int
    pass_out_of_drive_rate: int
    skip_pass_rate: int
    alley_oop_pass_rate: int
    gamble_for_steals: int
    help_defense_rate: int
    closeout_aggression: int
    box_out_rate: int
    usage_desire: int
    pace_preference: int
    foul_proneness: int
    shot_clock_tendency: int
    contested_shot_willingness: int


@dataclass
class CharacterTraits:
    leadership: int
    work_ethic: int
    clutch: int
    ego: int
    coachability: int
    temperament: int
    fan_favorite: int
    media_personality: int
    loyalty: int
    competitiveness: int


@dataclass
class DurabilityProfile:
    overall_durability: int
    ankle_health: int
    knee_health: int
    shoulder_health: int
    back_health: int
    wrist_hand_health: int
    foot_health: int
    concussion_risk: int
    soft_tissue_risk: int
    injury_history: list[InjuryRecord] = field(default_factory=list)


@dataclass
class PlayerStatus:
    health: str
    current_injury: ActiveInjury | None = None
    fatigue: float = 0.0
    morale: float = 1.0
    is_rookie: bool = False
    is_free_agent: bool = False
    is_restricted_fa: bool = False
    team_id: str | None = None


@dataclass
class Player:
    id: str
    bio: PlayerBio
    ratings: PlayerRatings
    shot_chart: ShotChartProfile
    tendencies: PlayerTendencies
    character: CharacterTraits
    durability: DurabilityProfile
    contract: Any = None
    status: PlayerStatus = field(default_factory=lambda: PlayerStatus(health="healthy"))
